text/0000-macro-naming.md

```diff
@@ -0,0 +1,165 @@
+named using the usual syntax. It is confusing that macros use their own system
+for modularisation. It would be far nicer if they were a more regular feature of
+Rust in this respect.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Defining macros
+
+This RFC does not propose changes to macro definitions. It is envisaged that
+definitions of procedural macros will change, see [this blog post](http://ncameron.org/blog/macro-plans-syntax/)
+for some rough ideas. I'm assuming that procedural macros will be defined in
+some function-like way and that these functions will be defined in modules in
+their own crate (to start with).
+
+Ordering of macro definitions in the source text will no longer be significant.
+A macro may be used before it is defined, as long as it can be named. That is,
+macros follow the same rules regarding ordering as other items. E.g., this will
+work:
+
+```
+foo!();
+
+macro_rules! foo { ... }
+```
+
+Macro expansion order is also not defined by source order. E.g., in `foo!(); bar!();`,
+`bar` may be expanded before `foo`. Ordering is only guaranteed as far as it is
+necessary. E.g., if `bar` is only defined by expanding `foo`, then `foo` must be
+expanded before `bar`.
+
+## Function-like macro uses
+
+A function-like macro use (c.f., attribute-like macro use) is a macro use which
+uses `foo!(...)` or `foo! ident (...)` syntax (where `()` may also be `[]` or `{}`).
+
+Macros may be named by using a `::`-separated path. Naming follows the same
+rules as other items in Rust.
+
+If a macro `baz` (by example or procedural) is defined in a module `bar` which
+is nested in `foo`, then it may be used anywhere in the crate using an
+absolute path: `::foo::bar::baz!(...)`. It can be used via relative paths in the
+usual way, e.g., inside `foo` as `bar::baz!()`.
+
+Macros declared inside a function body can only be used inside that function
+body.
+
+For procedural macros, the path must point to the function defining the macro.
+
+The grammar for macros is changed, anywhere we currently parser `name "!"`, we
+now parse `path "!"`. I don't think this introduces any issues.
+
+Name lookup follows the same name resolution rules as other items. See [RFC
+1560](https://github.com/rust-lang/rfcs/pull/1560) for details on how name
+resolution could be adapted to support this.
+
+## Attribute-like macro uses
+
+Attribute macros may also be named using a `::`-separated path. Other than
+appearing in an attribute, these also follow the usual Rust naming rules.
+
+E.g., `#[::foo::bar::baz(...)]` and `#[bar::baz(...)]` are uses of absolute and
+relative paths, respectively.
+
+
+## Importing macros
+
+Importing macros is done using `use` in the same way as other items. An `!` is
+not necessary in an import item. Macros are imported into their own namespace
+and do not shadow or overlap items with the same name in the type or value
+namespaces.
+
+E.g., `use foo::bar::baz;` imports the macro `baz` from the module `::foo::bar`.
```

Comment:
> Also I'm of the opinion my editor shouldn't know that something is a macro.

The point isn't to let editors know what is a macro is for any express purpose, per se, it's to let them understand the basic structure of the program without knowing any definitions so they can implement syntax highlighting and other features easily.

> it's very nice to be able to use macros to extend the language in a transparent way

I would argue that it's more transparent to tooling and humans when the macros are obviously marked at the invocation site. In fact, I might even prefer if the `!` *was* part of the macro name, but that would be inconsistent with existing syntax I mentioned before.